### PR TITLE
Allow login via TLS with externally hosted WebIDs

### PR DESCRIPTION
**Pull request by dmitrizagidulin**
(after verifying that the server is the preferred oidc issuer of that webid profile).

Addresses https://github.com/solid/node-solid-server/issues/521

### CODE REVIEW

File: lib/models/account-manager.js [hunk 1]
```diff
@@ -345,18 +345,19 @@ class AccountManager {
...
-    }
-
     if (!userConfig.webId && !userConfig.username) {
       throw new Error('Username or web id is required')
     }
 
+    if (userConfig.webId && !userConfig.username) {
+      userConfig.username = this.usernameFromWebId(userConfig.webId)
+    }
+
```
Comment (RubenVerborgh) [inline lib/models/account-manager.js]: @dmitrizagidulin MIght want to group those two `if`s in an `if (!userConfig.username)`?

File: lib/models/authenticator.js [hunk 2]
```diff
@@ -304,13 +310,27 @@ class TlsAuthenticator extends Authenticator {
    * @return {UserAccount}
    */
   ensureLocalUser (webId) {
-    if (this.accountManager.externalAccount(webId)) {
-      debug(`WebID URI ${JSON.stringify(webId)} is not a local account`)
+    const serverUri = this.accountManager.host.serverUri
 
-      throw new Error('Cannot login: Selected Web ID is not hosted on this server')
+    // if (this.accountManager.externalAccount(webId)) {
```
Comment (RubenVerborgh) [inline lib/models/authenticator.js]: @dmitrizagidulin Should this commented code still be here?

File: lib/models/authenticator.js [hunk 3]
```diff
@@ -304,13 +310,27 @@ class TlsAuthenticator extends Authenticator {
...
+    if (domainMatches(serverUri, webId)) {
+      // This is a locally hosted Web ID
+      return Promise.resolve(this.accountManager.userAccountFrom({ webId }))
     }
 
-    return this.accountManager.userAccountFrom({ webId })
+    debug(`WebID URI ${JSON.stringify(webId)} is not a local account, verifying preferred provider`)
+
+    return this.discoverProviderFor(webId)
+      .then(preferredProvider => {
```
Comment (RubenVerborgh) [inline lib/models/authenticator.js]: I wonder is "preferred" is too strong, given that we only have a predicate to indicate the provider (and we simply assume there is only one).

File: lib/models/authenticator.js (hunk 2 above)
Comment (dmitrizagidulin) [inline lib/models/authenticator.js]: Whoops, thanks

File: lib/models/authenticator.js (hunk 3 above)
Comment (dmitrizagidulin) [inline lib/models/authenticator.js]: Ah, true. I'll change it to 'authorized provider', then.

File: lib/models/account-manager.js (hunk 1 above)
Comment (dmitrizagidulin) [inline lib/models/account-manager.js]: Sure thing.

**Comment by dmitrizagidulin** [comment]
@RubenVerborgh updated. Let me know if there's anything else

**Comment by RubenVerborgh** [comment]
I've tested this with `bin/solid-test`  from #546 merged in.

### Questions beforehand
- What does `"auth": "oidc"` in my config still mean? Will it be both by default now?

### Logging in without an account (and without the right triple in my profile)
- I click "With certificate", and I receive "This server is not the authorized provider for Web ID https://ruben.verborgh.org/profile/#me". Perhaps this error message can be clarified with a solution? Such as "Please authorize add <x> <y> <z> to your profile." Then again, WebID-TLS will only be used by advanced users.

### Logging in without an account (but with the right triple)
- Logging in works. I get `Access denied for https://ruben.verborgh.org/profile/#me`, but that's because my data folder is empty.

So good to go for me, perhaps just the `"auth"` setting needs clarification.

**Comment by dmitrizagidulin** [comment]
> What does `"auth": "oidc"` in my config still mean? Will it be both by default now?

So, it's not quite both, by default. (I'll add the explanation below to the CHANGELOG upgrade notes).

`"auth": "tls"` means: "use the old WebID-TLS auth system", which means that the server prompts for a certificate on each request, etc. And the OIDC components are not initialized & not used.

`"auth": "oidc"` means: "use the WebID-OIDC auth system", and *optionally*, if the config entry `"localAuth": { "tls": <boolean> }` is not set to *false*, TLS will be offered as a *local login method*. However, the server will not prompt for certificates on each request, only on the TLS login api endpoint.

Does that make more sense? 

> This server is not the authorized provider error message

Yeah, agreed, that could be clarified. But like you said, it's for advanced users, there's basically gonna be like 4 users total who would want to use that.

**Comment by dmitrizagidulin** [comment]
What would be the best way to clarify the error message? Maybe link to the section in the spec about discovery from profile RDF?

**Comment by RubenVerborgh** [comment]
A link would probably be all that users need, indeed. Thanks!

**Pull request merged**